### DISCUSSION TITLE
h503 .cfg Files to enable Debug

### DISCUSSION THREAD

**Discussion by LightningStalker** [stm32duino/Arduino_Core_STM32#2383] [Q&A] [answered]
**Is your feature request/improvement related to a problem? Please describe.**

STM32H503RB Nucleo
analogWrite(PA5, xxval) is non-return
HAL_DAC_ConfigChannel() (called by analogWrite) is also non-return
Unable to trace further without debugging
No compile error but MCU "lockup"

**Describe the solution you'd like**

.cfg files and others related to openocd Debug for STM32H5xx

**Describe alternatives you've considered**

CubeIDE having its own issues

**Additional context**

.cfg files not found in stm32duino github
DAC config state unknown but doing conversion
HAL functions must be used (except ConfigChannel()
Nice MCU need to try more peripherals

**Comment by fpistm** [comment]
Hi @LightningStalker 

About H5 .cfg, it was not provided because openocd do not support it (yet?).
https://github.com/openocd-org/openocd/tree/master/tcl/target

If you want to debug, you can try using cmake and import the cmake project in STM32CubeIDE which include a customized openocd version.
Or you can change the platform.txt to use the openocd from the STM32CubeIDE and copy the .cfg in the variants/STM32H5xx/ folder.

Anyway this is advanced usage and do not support this, sorry.

Note:
Since Arduino IDE 2.3.x, the debug config has change and have to update it.

**Comment by LightningStalker** [comment]
After updates to stm32duino core still same behavior with HAL_DAC_ConfigChannel() on H503RB

Test code:

```
void setup() {
  pinMode(PA5, OUTPUT);

  // initialize serial communication at 9600 bits per second:
  Serial.begin(9600);
  analogReadResolution(12);
}

// the loop routine runs over and over again forever:
void loop() {
  static uint16_t dac_value = 0;

// CALL HAL_DAC_ConfigChannel():
  analogWrite(PA5, dac_value);
  dac_value = (dac_value < 4095) ? (16 + dac_value) : 0;
  // read the input on analog pin 0:
  int sensorValue = analogRead(A0);
  // print out the value you read:
  Serial.println(sensorValue);
  Serial.println(dac_value);
  delay(1);  // delay in between reads for stability
}
```

After some hours: 
```
[2024-07-24T06:21:03.734Z] SERVER CONSOLE DEBUG: onBackendConnect: gdb-server session connected. You can switch to "DEBUG CONSOLE" to see GDB interactions.
/opt/st/stm32cubeide_1.15.1/plugins/com.st.stm32cube.ide.mcu.externaltools.openocd.linux64_2.3.100.202312181736/tools/bin/openocd -c "gdb_port 50000" -c "tcl_port 50001" -c "telnet_port 50002" -s /home/robert/Arduino/h503_DAC -f /tmp/.mount_arduintiYITR/resources/app/plugins/cortex-debug/extension/support/openocd-helpers.tcl -f /home/robert/.arduino15/packages/STMicroelectronics/tools/xpack-openocd/0.12.0-3/openocd/scripts/interface/stlink.cfg -f /home/robert/.arduino15/packages/STMicroelectronics/hardware/stm32/2.8.1/debugger/select_hla.cfg -f /home/robert/.arduino15/packages/STMicroelectronics/tools/xpack-openocd/0.12.0-3/openocd/scripts/target/stm32h5x.cfg
Open On-Chip Debugger 0.12.0-00029-gf77e7cb03 (2023-12-15-13:24) [https://github.com/STMicroelectronics/OpenOCD]
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
CDRTOSConfigure
hla_swd
Error: The selected transport doesn't support this target
/home/robert/.arduino15/packages/STMicroelectronics/tools/xpack-openocd/0.12.0-3/openocd/scripts/target/stm32h5x.cfg:77: Error: 
in procedure 'script' 
at file "embedded:startup.tcl", line 28
at file "/home/robert/.arduino15/packages/STMicroelectronics/tools/xpack-openocd/0.12.0-3/openocd/scripts/target/stm32h5x.cfg", line 77
[2024-07-24T06:21:03.757Z] SERVER CONSOLE DEBUG: onBackendConnect: gdb-server session closed
GDB server session ended. This terminal will be reused, waiting for next session to start...
```

**Comment by LightningStalker** [comment]
stm32h5xx_hal_dac.c line 1355:
```
  if ((*(tmp_package) & 0x1FUL) == 0x0FUL)
```
Should evaluate FALSE but is instead TRUE bug maybe?

**Reply by fpistm** [reply]
This topic was for debug config ....

This line is part of the Cube H5 HAL.
If you think there is a bug on this line open an issue on the GitHub STMicroelectronics in the CubeH5 repository.

**Reply by LightningStalker** [reply]
Not sure if it is HAL bug or conflict between core setting.
Memory address test is new for me!  Abstraction normally.

**Reply by fpistm** [reply, marked as answer]
Hi @LightningStalker 

I've tested and met the issue too.
This is due to the ICACHE which is enabled preventing reading the PACKAGE version.
More info here:
https://community.st.com/t5/stm32-mcus/how-to-avoid-a-hardfault-when-icache-is-enabled-on-the-stm32h5/ta-p/630085

You can disable it adding a file named `hal_conf_extra.h` with:
```C
#define HAL_ICACHE_MODULE_DISABLED
```
